db/db_compaction_test.cc

```diff
@@ -5020,6 +5021,159 @@ TEST_F(DBCompactionTest, LevelPeriodicCompaction) {
+          if (if_restart) {
+            std::string* encoded_fieled = static_cast<std::string*>(arg);
+            *encoded_fieled = "";
+            PutVarint64(encoded_fieled, 0);
+          }
+        });
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+
+    // Just to add some extra random days to current time
+    Random rnd(test::RandomSeed());
+    int days = rnd.Uniform(100);
+
+    int periodic_compactions = 0;
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+        "LevelCompactionPicker::PickCompaction:Return", [&](void* arg) {
+          Compaction* compaction = reinterpret_cast<Compaction*>(arg);
+          auto compaction_reason = compaction->compaction_reason();
+          if (compaction_reason == CompactionReason::kPeriodicCompaction) {
+            periodic_compactions++;
+          }
+        });
+
+    // Case 1: Periodic Compaction during non-offpeak time
+    {
+      // Starting at 1:30PM
+      int now_hour = 13;
+      int now_minute = 30;
+      auto mock_clock =
+          std::make_shared<MockSystemClock>(env_->GetSystemClock());
+      auto mock_env = std::make_unique<CompositeEnvWrapper>(env_, mock_clock);
+      options.env = mock_env.get();
+      mock_clock->SetCurrentTime(days * kSecondsPerDay +
+                                 now_hour * kSecondsPerHour +
+                                 now_minute * kSecondsPerMinute);
+      // Offpeak is set from 12:30AM to 4:30AM
+      options.daily_offpeak_time_utc = "00:30-04:30";
+      Reopen(options);
+
+      for (int i = 0; i < kNumLevelFiles; ++i) {
+        for (int j = 0; j < kNumKeysPerFile; ++j) {
+          ASSERT_OK(
+              Put(Key(i * kNumKeysPerFile + j), rnd.RandomString(kValueSize)));
+        }
+        ASSERT_OK(Flush());
+      }
+      ASSERT_OK(dbfull()->TEST_WaitForCompact());
+      ASSERT_EQ("2", FilesPerLevel());
+      ASSERT_EQ(0, periodic_compactions);
+
+      // Move clock forward check if it triggers periodic comapaction
+      mock_clock->MockSleepForSeconds(5 * kSecondsPerDay + 1);
+      ASSERT_OK(Put("a", "1"));
+      ASSERT_OK(Flush());
+      ASSERT_OK(dbfull()->TEST_WaitForCompact());
+      // Assert that the files stay in the same level
+      ASSERT_EQ("3", FilesPerLevel());
+      // The two old files go through the periodic compaction process
+      ASSERT_EQ(2, periodic_compactions);
+      MoveFilesToLevel(1);
+      ASSERT_EQ("0,3", FilesPerLevel());
+
+      // Move forward 5 more days and more write
+      mock_clock->MockSleepForSeconds(5 * kSecondsPerDay + 1);
+      ASSERT_OK(Put("b", "2"));
+      if (if_restart) {
+        Reopen(options);
+      } else {
+        ASSERT_OK(Flush());
+      }
+      ASSERT_OK(dbfull()->TEST_WaitForCompact());
+      ASSERT_EQ("1,3", FilesPerLevel());
+      ASSERT_EQ(5, periodic_compactions);
+
+      Destroy(options);
+    }
+    periodic_compactions = 0;
+    // Case 2: Periodic Compaction during off-peak.
```

Comment:
I'm not sure why this is separate and somewhat different from case 1. AFAIK, we are primarily interested in testing a difference in behavior between having daily_offpeak_time_utc enabled or not, or between different settings of daily_offpeak_time_utc. This test doesn't do that. For the best control against confounding variables, you should have one test script that runs with the different daily_offpeak_time_utc settings and expects different results. Recommend a loop with SCOPED_TRACE.